# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/StaticRange.cpp
# Commit: 0837e15babab
# Full Hash: 0837e15bababe485cb399b462edcfccd49935126
# Author: Sean Feng <sefeng@mozilla.com>
# Date: 2024-03-23 09:29:17
# Regressor Bug: 1867058
# File Overlap Count: 1
# Description:
#   Bug 1867058 - Part 2: Make both StaticRange and nsRange to have a way to cross the trees r=jjaschke,smaug,dom-core
#   
#   For StaticRange, simply allows node to cross the trees.
#   
#   For nsRange, when nodes are in different trees, in addition
# ==============================================================================

diff -r 8ce7972ea4df -r 0837e15babab dom/base/StaticRange.cpp
--- a/dom/base/StaticRange.cpp	Fri Mar 22 17:05:48 2024 +0000
+++ b/dom/base/StaticRange.cpp	Fri Mar 22 17:05:48 2024 +0000
@@ -100,6 +100,13 @@
     return false;
   }
 
+  MOZ_ASSERT(mAreStartAndEndInSameTree ==
+             (RangeUtils::ComputeRootNode(mStart.Container()) ==
+              RangeUtils::ComputeRootNode(mEnd.Container())));
+  if (!mAreStartAndEndInSameTree) {
+    return false;
+  }
+
   const Maybe<int32_t> pointOrder = nsContentUtils::ComparePoints(mStart, mEnd);
   return pointOrder.isSome() && *pointOrder <= 0;
 }
@@ -119,6 +126,9 @@
   if (checkCommonAncestor) {
     UpdateCommonAncestorIfNecessary();
   }
+
+  mAreStartAndEndInSameTree = RangeUtils::ComputeRootNode(mStart.Container()) ==
+                              RangeUtils::ComputeRootNode(mEnd.Container());
 }
 
 /* static */